### PR TITLE
SDE GAN example

### PR DESCRIPTION
**Pull request by patrick-kidger**
Added an SDE-GAN example. Only a draft as I'm not satisfied that it's working yet / that it's as fast as it could be.

I've added a couple other things in here too. For one, I've added a CHANGELOG.txt file. (Anything I've left out?)
I've also tweaked the citation request with the new paper, to reflect the huge amount of work we've put into this repo for that paper.

### CODE REVIEW

**Comment by lxuechen** [comment]
Thanks for the update. Is the changelog necessary, as I've been documenting things in the release notes? I'll take a look at the other stuff tomorrow.

**Comment by patrick-kidger** [comment]
Haha, I honestly hadn't spotted the release notes. Agreed, the changelog is unnecessary.

FYI this example still doesn't work yet. Still needs more tweaks. (The same was true when we wrote the paper, it took me like 2 months of tweaking to get it working. Hopefully we can do it quicker now, but this isn't surprising.)

**Comment by lxuechen** [comment]
Thanks for the update, @patrick-kidger ! I'm assuming the example works now. 

I actually have further thoughts about this gan model, though I'm quite busy before the xmax break and couldn't really afford to context switch between different things. 

Would you mind if I go through this PR in a week of time? 

Aside, build for Py3.6 seems to be failing due to numpy being incompatible after its latest release.

**Comment by patrick-kidger** [comment]
Yup, the example works now! The most important points were to change the initialisation and to switch from Adam to SGD.

I'd be very interested to hear your thoughts on SDEs/GANs. After xmas is fine; likewise no rush on this PR. It looks like SDE-GANs are probably going to be in ICML rather than ICLR, after all...

Aside, eurgh. Looks like the problem is that it's grabbing the SciPy pre-release, which [requires Python 3.7+](http://scipy.github.io/devdocs/release.1.6.0.html).

**Comment by lxuechen** [comment]
Not related to this thread in particular, but I found [this](https://pytorch.org/docs/master/generated/torch.vmap.html) during late night reading. Some food for thought regarding the log-ODE scheme.

**Comment by patrick-kidger** [comment]
Ah, thankyou! I've actually come across this before. Unfortunately I don't think it composes with `autograd.grad` in the desired way though -- from a private discussion with one of PyTorch devs: "The current vmap prototype unfortunately does not support doing a batch vjp where the input and the v vector share a batch dimension."

At some point I expect we'll revisit the log-ODE method, as I believe James wants to write a paper on it. When that happens one option would be to write a for loop over the vjp, in C++ parallelised with OpenMP.

File: examples/sde_gan.py [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+        # The +1 corresponds to time. When solving CDEs, It turns out to be most natural to treat time as just another
+        # channel: in particular this makes handling irregular data quite easy, when the times may be different between
+        # different samples in the batch.
+
+        Y = torchcde.LinearInterpolation(ys_coeffs)
+        Y0 = Y.evaluate(Y.interval[0])
+
+        # TODO: step size
+        h0 = self._initial(Y0)
+        # We happen to use midpoint with step_size=1e-1 for consistency with the SDE solve, but that isn't important.
```
Comment (lxuechen) [inline examples/sde_gan.py:183]: Does this refer to the step size for generating the data, or that for the generator solve?

**Review by lxuechen — COMMENTED** [review]
Thanks for the example! I've only read the code but have not played with the code, so these mostly are just questions. Are there ways to simplify this example and potentially speed up running it (e.g. by reducing the time horizon)?

File: examples/sde_gan.py [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+
+    # Data
+    print("Generating data...")
+    ts, ys = get_data()
+    print("Generated data.")
+    data_size = ys.size(-1) - 1  # How many channels the data has (not including time, hence the minus one).
+    ts = ts.to(device)
+    ys_coeffs = torchcde.linear_interpolation_coeffs(ys)  # as per neural CDEs.
+    dataset = torch.utils.data.TensorDataset(ys_coeffs)
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, num_workers=6, pin_memory=is_cuda)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply examples/sde_gan.py:334]: Generically, yes. (It's my go-to choice for DataLoaders.) I've not tested it on this particular problem.
In any case this dataset is actually small enough that it entirely fits on the GPU, so I've modified this now.

File: examples/sde_gan.py [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+
+    def forward(self, ts, batch_size):
+        # ts has shape (t_size,) and corresponds to the points we want to evaluate the SDE at.
+
+        ###################
+        # Actually solve the SDE.
+        ###################
+        init_noise = torch.randn(batch_size, self._initial_noise_size, device=ts.device)
+        x0 = self._initial(init_noise)
+        # TODO: step size
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply examples/sde_gan.py:125]: It's just slower. The step size used for generating the data needn't be connected to the step size we use in our model. (We don't output the data every step; only every 10 steps.)

This actually touches on what I've found to be the single biggest issue with neural SDEs: training them is really, really slow. Even with this example, which:
- has a large batch size;
- has a huge step size;
- isn't making adaptive steps;
- is backprop'ing through the solver, without the adjoint method;
- is on a decent GPU;
- has the benefit of having been tweaked for a few months now.
Even this still takes a substantial amount of time to run.

I'm not really sure what the solution is, to be honest. (Lots of GPUs would be one way.) Large step sizes are just one way of ameliorating this -- I suspect the prior on model space is pretty similar to that of small step sizes.

File: examples/sde_gan.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+    # What we're doing here is very crude -- one can definitely imagine smarter ways of doing things.
+    # (e.g. pretraining the t=0 distribution)
+    with torch.no_grad():
+        for param in generator._initial.parameters():
+            param *= init_mult1
+        for param in generator._func.parameters():
+            param *= init_mult2
+
+    # Optimisers. Simple SGD is a good choice here.
+    generator_optimiser = torch.optim.SGD(generator.parameters(), lr=lr, weight_decay=weight_decay)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply examples/sde_gan.py:357]: See below.

File: examples/sde_gan.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+    ys_coeffs = torchcde.linear_interpolation_coeffs(ys)  # as per neural CDEs.
+    dataset = torch.utils.data.TensorDataset(ys_coeffs)
+    dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, num_workers=6, pin_memory=is_cuda)
+
+    # Models
+    generator = Generator(data_size, initial_noise_size, noise_size, hidden_size, mlp_size, num_layers).to(device)
+    discriminator = Discriminator(data_size, hidden_size, mlp_size, num_layers).to(device)
+    # Weight averaging really helps with GAN training.
+    # (Could probably also substitute this for something like negative momentum.)
+    averaged_generator = swa_utils.AveragedModel(generator)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply examples/sde_gan.py:341]: See below.

File: examples/sde_gan.py [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+
+            batch_size = real_samples.size(0)
+            total_samples += batch_size
+            total_loss += loss.item() * batch_size
+    return total_loss / total_samples
+
+
+def main():
+    # Architectural hyperparameters. These are quite small for illustrative purposes.
+    initial_noise_size = 5  # How many noise dimensions to sample at the start of the SDE.
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply examples/sde_gan.py:300]: I'm not a fan of argparse in example code: increased boilerplate; reduced readability.

File: examples/sde_gan.py [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+
+class Discriminator(torch.nn.Module):
+    def __init__(self, data_size, hidden_size, mlp_size, num_layers):
+        super(Discriminator, self).__init__()
+        self._initial = MLP(1 + data_size, hidden_size, mlp_size, num_layers, tanh=False)
+        self._func = DiscriminatorFunc(data_size, hidden_size, mlp_size, num_layers)
+        self._readout = torch.nn.Linear(hidden_size, 1)
+
+        # Final layer has an easier problem to solve, so increase its learning rate.
+        self._readout.weight.register_hook(lambda grad: 100 * grad)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply examples/sde_gan.py:169]: This is just a standard trick for neural CDEs. Again, seems to help.
(Good point about Adam, this way of implementing increased LR needs some care.)

File: examples/sde_gan.py (hunk 1 above)
Comment (patrick-kidger) [reply examples/sde_gan.py:183]: I've removed this now.

**Comment by patrick-kidger** [comment]
A fair few of your comments above are about the optimisation procedure. This is easily the thing that actually seems to matter most.

I didn't try every combination of SGD/Adam/Adadelta/... with every combination of weight decay, SWA, etc, but I'm actually very interested in what the optimal way of doing this is.

As a general summary of my observations:
- Adam/AMSgrad give losses that jump around all over the show during training, whilst Adadelta gives very well-behaved smoothly-varying losses. However only SGD manages to get anything approaching good convergence towards the end of training, though: the others tend to get stuck at some loss away from zero. (Probably related to the theoretical lack-of-convergence that these optimisers exhibit.)
- Weight decay seems to help a little bit.
- SWA can save you from the oscillating losses that Adam gives you, helps a fair chunk with SGD, but does basically nothing with Adadelta. I've only tried uniform weight averaging, not EMA.

Judging from the GAN optimisation literature at the moment, it seems like no-one else has much idea either. SWA is obviously related to the stronger convergence guarantees that Cesaro means often get / is a trick to avoid the diffusive regime at the end of training. The tricks used in Ada-* help at the start of training but break things at the end of training. Weight decay and negative momentum seem related to the idea of solving a GAN as an ODE, but that's almost certainly suboptimal for min-max games, just like it's suboptimal for minimisation problems.

All in all things seem to be in a bit of a mess. It feels to me like all the basic pieces are there in the literature, and someone just needs to figure out the correct things to put together to get it working reliably.

As it stands this example represents a "good enough" trade-off between readability / efficacy / use of only standard library without custom optimisers.

Things I'd like to try, but decided were out of scope for an intentionally simple example:
- Negative momentum. Unfortunately the default PyTorch optimizers through an error if you try it.
- Adabound, which interpolates between Adam and SGD. Something interpolating between Adadelta and SGD would also be interesting.
- Using Adadelta to begin training and then switch to SGD at the end.
- Coming up with some custom optimiser that has the right blend of the above ideas!

File: examples/sde_gan.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhirschfeld) [reply examples/sde_gan.py:294]: If you want a CLI but are worried about boilerplate maybe https://github.com/google/python-fire might fit the bill?
(Haven't tried it myself so YMMV)

**Comment by patrick-kidger** [comment]
FYI even though I think this PR is done, I'm thinking of leaving it unmerged until the paper is on arXiv (so we have a better link than just the workshop paper).

File: examples/sde_gan.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lxuechen) [reply examples/sde_gan.py:294]: I have been using fire for quite a while, though I was relunctant to include it for other examples, as it required another dependency. 

Thinking about it now, it's not that big a hassle, and I have modified the other examples like so. 

In general not a huge deal though.

[the remark this answers is not in the document]
Comment (lxuechen) [reply examples/sde_gan.py:294]: A side comment is that I would still recommend using some method so that the script accepts command line args directly. Or perhaps at the very least, make args like `initial_noise_size` parameters of the function so they can be tweaked if people wanted to.

**Comment by patrick-kidger** [comment]
Right, finally done with this PR and ready to merge it, I think.

**Comment by lxuechen** [comment]
> Finally done with this PR and ready to merge it, I think. @lxuechen how does it look?

Looks good. If you could rebase from master to rid the test failures, then I think it's mostly ready for merge; otherwise I can try do the rebase some time over the weekend. Overall, the example seems a somewhat more complicated than the others, but understandably training GANs isn't easy. 

Sorry again for this ultra slow response.

**Comment by patrick-kidger** [comment]
No worries at all.

The test failure looks to be because the version number isn't updated. I can bump the number if you really want, but as we're only touching the README and a new example file then I think it should be fine to just merge anyway.

I've just done a small update to start using Fire, and add the license at the top.

**Review by lxuechen — APPROVED** [review]
